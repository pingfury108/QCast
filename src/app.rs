use async_trait::async_trait;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::Queue,
    boot::{create_app, BootResult, StartMode},
    config::Config,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    task::Tasks,
    Result,
};
use migration::Migrator;
use std::path::Path;

#[allow(unused_imports)]
use crate::{controllers, initializers, models::_entities::users, tasks, workers};

pub struct App;
#[async_trait]
impl Hooks for App {
    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA")
                .or(option_env!("GITHUB_SHA"))
                .unwrap_or("dev")
        )
    }

    async fn boot(
        mode: StartMode,
        environment: &Environment,
        config: Config,
    ) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment, config).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .add_route(controllers::books::routes())
            .add_route(controllers::chapters::routes())
            .add_route(controllers::medias::routes())
            .add_route(controllers::public::routes())
            .add_route(controllers::auth::routes())
            .add_route(controllers::dashboard::routes())
            // 后台管理路由
            .add_route(controllers::admin::users::routes())
            .add_route(controllers::admin::groups::routes())
    }
    async fn connect_workers(_ctx: &AppContext, _queue: &Queue) -> Result<()> {
        // 暂无 worker
        Ok(())
    }

    #[allow(unused_variables)]
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::create_superadmin::CreateSuperadmin);
        // tasks-inject (do not remove)
    }
    async fn truncate(ctx: &AppContext) -> Result<()> {
        truncate_table(&ctx.db, users::Entity).await?;
        Ok(())
    }
    async fn seed(ctx: &AppContext, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(&ctx.db, &base.join("users.yaml").display().to_string())
            .await?;
        Ok(())
    }
}
