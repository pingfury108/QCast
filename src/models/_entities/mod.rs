//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub mod prelude;

pub mod books;
pub mod chapters;
pub mod medias;
pub mod site_settings;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
