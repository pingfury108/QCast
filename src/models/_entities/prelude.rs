//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub use super::books::Entity as Books;
pub use super::chapters::Entity as Chapters;
pub use super::medias::Entity as Medias;
pub use super::site_settings::Entity as SiteSettings;
pub use super::user_group_members::Entity as UserGroupMembers;
pub use super::user_groups::Entity as UserGroups;
pub use super::users::Entity as Users;
