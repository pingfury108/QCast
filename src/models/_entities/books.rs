//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "books")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub title: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
    pub user_id: i32,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
