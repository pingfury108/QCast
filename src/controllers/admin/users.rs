use crate::models::_entities::users as users_entity;
use crate::models::users;
use axum::extract::Query;
use axum::routing::method_routing::delete as axum_delete;
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct ListParams {
    pub page: Option<u64>,
    pub per_page: Option<u64>,
    pub search: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateRoleParams {
    pub is_staff: bool,
    pub is_superuser: bool,
}

#[derive(Debug, Serialize)]
pub struct UserListResponse {
    pub users: Vec<users::Model>,
    pub pagination: PaginationInfo,
}

#[derive(Debug, Serialize)]
pub struct PaginationInfo {
    pub page: u64,
    pub per_page: u64,
    pub total_pages: u64,
}

/// 列出所有用户（需要管理员权限）
pub async fn list(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Query(params): Query<ListParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    if !user.is_admin() {
        return unauthorized("需要管理员权限");
    }

    let page = params.page.unwrap_or(1);
    let per_page = params.per_page.unwrap_or(20);

    let (user_list, total_pages) = if let Some(keyword) = params.search {
        users::Model::search(&ctx.db, &keyword, page, per_page).await?
    } else {
        users::Model::list_all(&ctx.db, page, per_page).await?
    };

    format::json(UserListResponse {
        users: user_list,
        pagination: PaginationInfo {
            page,
            per_page,
            total_pages,
        },
    })
}

/// 更新用户角色（需要超级管理员权限）
pub async fn update_role(
    auth: auth::JWT,
    Path(user_id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<UpdateRoleParams>,
) -> Result<Response> {
    let admin = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    if !admin.is_super_admin() {
        return unauthorized("需要超级管理员权限");
    }

    let updated_user =
        users::Model::update_admin_status(&ctx.db, user_id, params.is_staff, params.is_superuser)
            .await?;

    format::json(updated_user)
}

/// 删除用户（需要超级管理员权限）
pub async fn delete(
    auth: auth::JWT,
    Path(user_id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let admin = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    if !admin.is_super_admin() {
        return unauthorized("需要超级管理员权限");
    }

    // 不允许删除自己
    if admin.id == user_id {
        return bad_request("不能删除自己");
    }

    users_entity::Entity::delete_by_id(user_id)
        .exec(&ctx.db)
        .await?;

    format::empty()
}

/// 获取统计信息
pub async fn stats(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    if !user.is_admin() {
        return unauthorized("需要管理员权限");
    }

    let total_users = users::Model::count_all(&ctx.db).await?;
    let total_admins = users::Model::count_admins(&ctx.db).await?;

    format::json(serde_json::json!({
        "total_users": total_users,
        "total_admins": total_admins,
    }))
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/api/admin/users")
        .add("/", get(list))
        .add("/stats", get(stats))
        .add("/{id}/role", put(update_role))
        .add("/{id}", axum_delete(delete))
}
