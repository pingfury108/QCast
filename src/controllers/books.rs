#![allow(clippy::missing_errors_doc)]
#![allow(clippy::unnecessary_struct_initialization)]
#![allow(clippy::unused_async)]
use axum::debug_handler;
use axum::extract::Query;
use axum::routing::method_routing::delete as axum_delete;
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};

use crate::models::_entities::books::{ActiveModel, Column, Entity, Model};
use crate::models::users;
use crate::views::books::{BookResponse, BookTreeResponse};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateParams {
    pub title: String,
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateParams {
    pub title: Option<String>,
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ReorderParams {
    pub sort_order: i32,
}

async fn load_item(ctx: &AppContext, id: i32, user_id: i32) -> Result<Model> {
    let item = Entity::find_by_id(id)
        .filter(Column::UserId.eq(user_id))
        .one(&ctx.db)
        .await?;
    item.ok_or_else(|| Error::NotFound)
}

/// 获取当前用户的所有书籍
#[debug_handler]
pub async fn list(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    let books = Entity::find()
        .filter(Column::UserId.eq(user.id))
        .all(&ctx.db)
        .await?;

    let responses: Vec<BookResponse> = books.into_iter().map(BookResponse::from).collect();

    format::json(responses)
}

/// 创建书籍
#[debug_handler]
pub async fn create(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Json(params): Json<CreateParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    let item = ActiveModel {
        user_id: Set(user.id),
        title: Set(params.title),
        description: Set(params.description),
        cover_image: Set(params.cover_image),
        parent_id: Set(params.parent_id),
        sort_order: Set(params.sort_order),
        is_public: Set(params.is_public),
        ..Default::default()
    };

    let item = item.insert(&ctx.db).await?;

    format::json(BookResponse::from(item))
}

/// 获取书籍详情
#[debug_handler]
pub async fn show(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    format::json(BookResponse::from(item))
}

/// 更新书籍
#[debug_handler]
pub async fn update(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<UpdateParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    let mut item = item.into_active_model();

    if let Some(title) = params.title {
        item.title = Set(title);
    }
    if let Some(description) = params.description {
        item.description = Set(Some(description));
    }
    if let Some(cover_image) = params.cover_image {
        item.cover_image = Set(Some(cover_image));
    }
    if let Some(parent_id) = params.parent_id {
        item.parent_id = Set(Some(parent_id));
    }
    if let Some(sort_order) = params.sort_order {
        item.sort_order = Set(Some(sort_order));
    }
    if let Some(is_public) = params.is_public {
        item.is_public = Set(Some(is_public));
    }

    let item = item.update(&ctx.db).await?;

    format::json(BookResponse::from(item))
}

/// 删除书籍
#[debug_handler]
pub async fn delete(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    item.delete(&ctx.db).await?;
    format::empty()
}

/// 搜索书籍
#[debug_handler]
pub async fn search(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Query(params): Query<std::collections::HashMap<String, String>>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    let query = params.get("q").map(|q| q.as_str()).unwrap_or("");

    if query.is_empty() {
        return format::json(Vec::<BookResponse>::new());
    }

    let books = Model::search(&ctx.db, user.id, query).await?;
    let responses: Vec<BookResponse> = books.into_iter().map(BookResponse::from).collect();

    format::json(responses)
}

/// 获取书籍的完整树形结构
#[debug_handler]
pub async fn tree(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _ = load_item(&ctx, id, user.id).await?;

    let book_tree = Model::get_tree(&ctx.db, id).await?;
    let tree_response = BookTreeResponse::from(book_tree);
    format::json(tree_response)
}

/// 调整书籍顺序
#[debug_handler]
pub async fn reorder(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<ReorderParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    let mut item = item.into_active_model();
    item.sort_order = Set(Some(params.sort_order));
    let item = item.update(&ctx.db).await?;

    format::json(BookResponse::from(item))
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("api/books/")
        .add("/", get(list))
        .add("/", post(create))
        .add("/search", get(search))
        .add("/{id}", get(show))
        .add("/{id}", put(update))
        .add("/{id}", patch(update))
        .add("/{id}", axum_delete(delete))
        .add("/{id}/tree", get(tree))
        .add("/{id}/reorder", post(reorder))
}
