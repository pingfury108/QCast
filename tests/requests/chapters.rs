use insta::{assert_debug_snapshot, with_settings};
use loco_rs::app::AppContext;
use loco_rs::testing::prelude::*;
use loco_rs::TestServer;
use qcast::app::App;
use qcast::models::_entities::books;
use qcast::models::users;
use qcast::views::chapters::ChapterResponse;
use sea_orm::EntityTrait;
use serde_json::json;
use serial_test::serial;

use super::prepare_data::{auth_header, init_user_login};

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("chapters_request");
        let _guard = settings.bind_to_scope();
    };
}

fn cleanup_chapter_response() -> Vec<(&'static str, &'static str)> {
    vec![
        (r#""id":\s*\d+"#, r#""id": ID"#),
        (r#""book_id":\s*\d+"#, r#""book_id": BOOK_ID"#),
        (r#""sort_order":\s*\d+"#, r#""sort_order": ORDER"#),
        (r#""created_at":\s*"[^"]+""#, r#""created_at": "DATE""#),
        (r#""updated_at":\s*"[^"]+""#, r#""updated_at": "DATE""#),
    ]
}

/// 创建测试书籍
async fn create_test_book(
    request: &TestServer,
    ctx: &AppContext,
    auth_header: &(axum::http::HeaderName, axum::http::HeaderValue),
) -> books::Model {
    let payload = json!({
        "title": "Test Book for Chapters",
        "description": "A test book"
    });

    let response = request
        .post("/api/books")
        .add_header(auth_header.0.clone(), auth_header.1.clone())
        .json(&payload)
        .await;

    assert_eq!(response.status_code(), 200);

    let response_text = response.text();
    let book_response: serde_json::Value = serde_json::from_str(&response_text).unwrap();

    books::Entity::find_by_id(book_response["id"].as_i64().unwrap() as i32)
        .one(&ctx.db)
        .await
        .unwrap()
        .unwrap()
}

/// 创建测试章节
async fn create_test_chapter(
    request: &TestServer,
    book_id: i32,
    auth_header: &(axum::http::HeaderName, axum::http::HeaderValue),
    title: &str,
    description: Option<&str>,
) -> ChapterResponse {
    let payload = if let Some(desc) = description {
        json!({
            "title": title,
            "description": desc
        })
    } else {
        json!({
            "title": title
        })
    };

    let response = request
        .post(&format!("/api/books/{}/chapters", book_id))
        .add_header(auth_header.0.clone(), auth_header.1.clone())
        .json(&payload)
        .await;

    assert_eq!(response.status_code(), 200);

    let response_text = response.text();
    serde_json::from_str(&response_text).unwrap()
}

#[tokio::test]
#[serial]
async fn can_create_chapter() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        let payload = json!({
            "title": "Chapter 1",
            "description": "First chapter"
        });

        let response = request
            .post(&format!("/api/books/{}/chapters", book.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        assert_eq!(chapter["title"], "Chapter 1");
        assert_eq!(chapter["description"], "First chapter");
        assert_eq!(chapter["sort_order"], 1);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_create_chapter_with_sort_order() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        let payload = json!({
            "title": "Chapter 5",
            "description": "Fifth chapter",
            "sort_order": 5
        });

        let response = request
            .post(&format!("/api/books/{}/chapters", book.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        assert_eq!(chapter["title"], "Chapter 5");
        assert_eq!(chapter["sort_order"], 5);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_list_chapters() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        // 创建多个章节
        create_test_chapter(
            &request,
            book.id,
            &auth_header,
            "Chapter 1",
            Some("First chapter"),
        )
        .await;
        create_test_chapter(
            &request,
            book.id,
            &auth_header,
            "Chapter 2",
            Some("Second chapter"),
        )
        .await;
        create_test_chapter(&request, book.id, &auth_header, "Chapter 3", None).await;

        let response = request
            .get(&format!("/api/books/{}/chapters", book.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let chapters: Vec<serde_json::Value> = serde_json::from_str(&response_text).unwrap();

        assert_eq!(chapters.len(), 3);
        assert_eq!(chapters[0]["title"], "Chapter 1");
        assert_eq!(chapters[1]["title"], "Chapter 2");
        assert_eq!(chapters[2]["title"], "Chapter 3");

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(chapters);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_chapter_by_id() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;
        let chapter = create_test_chapter(
            &request,
            book.id,
            &auth_header,
            "Test Chapter",
            Some("Test Description"),
        )
        .await;

        let response = request
            .get(&format!("/api/books/{}/chapters/{}", book.id, chapter.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let response_chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        assert_eq!(response_chapter["title"], "Test Chapter");
        assert_eq!(response_chapter["description"], "Test Description");

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(response_chapter);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_update_chapter() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;
        let chapter = create_test_chapter(
            &request,
            book.id,
            &auth_header,
            "Original Title",
            Some("Original Description"),
        )
        .await;

        let payload = json!({
            "title": "Updated Title",
            "description": "Updated Description",
            "sort_order": 10
        });

        let response = request
            .put(&format!("/api/books/{}/chapters/{}", book.id, chapter.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let updated_chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        assert_eq!(updated_chapter["title"], "Updated Title");
        assert_eq!(updated_chapter["description"], "Updated Description");
        assert_eq!(updated_chapter["sort_order"], 10);

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(updated_chapter);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_delete_chapter() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;
        let chapter =
            create_test_chapter(&request, book.id, &auth_header, "Chapter to Delete", None).await;

        // 删除章节
        let response = request
            .delete(&format!("/api/books/{}/chapters/{}", book.id, chapter.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .await;

        assert_eq!(response.status_code(), 200);

        // 验证章节已删除
        let response = request
            .get(&format!("/api/books/{}/chapters/{}", book.id, chapter.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .await;

        assert_eq!(response.status_code(), 404);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_reorder_chapter() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;
        let chapter =
            create_test_chapter(&request, book.id, &auth_header, "Chapter to Reorder", None).await;

        let payload = json!({
            "sort_order": 5
        });

        let response = request
            .post(&format!(
                "/api/books/{}/chapters/{}/reorder",
                book.id, chapter.id
            ))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let reordered_chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        assert_eq!(reordered_chapter["sort_order"], 5);

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(reordered_chapter);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_batch_reorder_chapters() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        // 创建多个章节
        let chapter1 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 1", None).await;
        let chapter2 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 2", None).await;
        let chapter3 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 3", None).await;

        let payload = json!({
            "chapter_ids": [chapter3.id, chapter1.id, chapter2.id]
        });

        let response = request
            .post(&format!("/api/books/{}/chapters/batch-reorder", book.id))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let chapters: Vec<serde_json::Value> = serde_json::from_str(&response_text).unwrap();

        assert_eq!(chapters.len(), 3);
        assert_eq!(chapters[0]["id"], chapter3.id);
        assert_eq!(chapters[0]["sort_order"], 1);
        assert_eq!(chapters[1]["id"], chapter1.id);
        assert_eq!(chapters[1]["sort_order"], 2);
        assert_eq!(chapters[2]["id"], chapter2.id);
        assert_eq!(chapters[2]["sort_order"], 3);

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(chapters);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_move_chapter_up() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        // 创建多个章节
        let chapter1 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 1", None).await;
        let chapter2 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 2", None).await;

        // 手动设置 chapter2 的顺序为 2
        let payload = json!({
            "sort_order": 2
        });

        let response = request
            .post(&format!(
                "/api/books/{}/chapters/{}/reorder",
                book.id, chapter2.id
            ))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        // 现在尝试将 chapter2 上移
        let response = request
            .post(&format!(
                "/api/books/{}/chapters/{}/move-up",
                book.id, chapter2.id
            ))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let moved_chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        // chapter2 现在应该在位置 1
        assert_eq!(moved_chapter["id"], chapter2.id);
        assert_eq!(moved_chapter["sort_order"], 1);

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(moved_chapter);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_move_chapter_down() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        // 创建多个章节
        let chapter1 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 1", None).await;
        let chapter2 =
            create_test_chapter(&request, book.id, &auth_header, "Chapter 2", None).await;

        // 尝试将 chapter1 下移
        let response = request
            .post(&format!(
                "/api/books/{}/chapters/{}/move-down",
                book.id, chapter1.id
            ))
            .add_header(auth_header.0.clone(), auth_header.1.clone())
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        let moved_chapter: serde_json::Value = serde_json::from_str(&response_text).unwrap();

        // chapter1 现在应该在位置 2
        assert_eq!(moved_chapter["id"], chapter1.id);
        assert_eq!(moved_chapter["sort_order"], 2);

        with_settings!({filters => cleanup_chapter_response()}, {
            assert_debug_snapshot!(moved_chapter);
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn requires_authentication() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        let logged_in_user = init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = auth_header(&logged_in_user.token);
        let auth_header = (auth_key, auth_value);

        let book = create_test_book(&request, &ctx, &auth_header).await;

        // 尝试不带认证访问
        let response = request
            .get(&format!("/api/books/{}/chapters", book.id))
            .await;

        // 应该返回 401 Unauthorized
        assert_eq!(response.status_code(), 401);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn cannot_access_other_user_chapter() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        // 创建第一个用户
        let user1 = init_user_login(&request, &ctx).await;
        let (auth_key1, auth_value1) = auth_header(&user1.token);
        let auth_header1 = (auth_key1, auth_value1);

        let book1 = create_test_book(&request, &ctx, &auth_header1).await;
        let chapter = create_test_chapter(
            &request,
            book1.id,
            &auth_header1,
            "Secret Chapter",
            Some("Secret content"),
        )
        .await;

        // 创建第二个用户
        let user2_payload = json!({
            "name": "User 2",
            "email": "user2@test.com",
            "password": "password123"
        });

        let response = request
            .post("/api/auth/register")
            .json(&user2_payload)
            .await;
        assert_eq!(response.status_code(), 200);

        let user2 = users::Model::find_by_email(&ctx.db, "user2@test.com")
            .await
            .unwrap();

        let verify_payload = json!({
            "token": user2.email_verification_token,
        });

        request.post("/api/auth/verify").json(&verify_payload).await;

        let login_response = request
            .post("/api/auth/login")
            .json(&serde_json::json!({
                "email": "user2@test.com",
                "password": "password123"
            }))
            .await;

        let login_response_text = login_response.text();
        let user2_login: serde_json::Value = serde_json::from_str(&login_response_text).unwrap();
        let user2_token = user2_login["token"].as_str().unwrap();
        let (auth_key2, auth_value2) = auth_header(user2_token);

        // 用户2尝试访问用户1的章节
        let response = request
            .get(&format!("/api/books/{}/chapters/{}", book1.id, chapter.id))
            .add_header(auth_key2.clone(), auth_value2.clone())
            .await;

        // 应该返回 404 Not Found（因为用户2无法访问用户1的书籍）
        assert_eq!(response.status_code(), 404);
    })
    .await;
}
