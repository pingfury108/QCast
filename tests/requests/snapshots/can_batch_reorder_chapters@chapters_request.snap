---
source: tests/requests/chapters.rs
expression: chapters
---
[
    Object {
        "book_id": Number(1),
        "created_at": String("2025-10-08T15:50:34Z"),
        "description": Null,
        "id": Number(3),
        "level": Number(0),
        "media_count": Number(0),
        "parent_id": Null,
        "path": String("3"),
        "sort_order": Number(1),
        "title": String("Chapter 3"),
        "updated_at": String("2025-10-08T15:50:34.284248Z"),
    },
    Object {
        "book_id": Number(1),
        "created_at": String("2025-10-08T15:50:34Z"),
        "description": Null,
        "id": Number(1),
        "level": Number(0),
        "media_count": Number(0),
        "parent_id": Null,
        "path": String("1"),
        "sort_order": Number(2),
        "title": String("Chapter 1"),
        "updated_at": String("2025-10-08T15:50:34.284670Z"),
    },
    Object {
        "book_id": Number(1),
        "created_at": String("2025-10-08T15:50:34Z"),
        "description": Null,
        "id": Number(2),
        "level": Number(0),
        "media_count": Number(0),
        "parent_id": Null,
        "path": String("2"),
        "sort_order": Number(3),
        "title": String("Chapter 2"),
        "updated_at": String("2025-10-08T15:50:34.285002Z"),
    },
]
